lib/Demangling/Demangler.cpp

```diff
@@ -1532,34 +1532,53 @@ NodePointer Demangler::demangleExtensionContext() {
 
 /// Associate any \c OpaqueReturnType nodes with the declaration whose opaque
 /// return type they refer back to.
-static Node *setParentForOpaqueReturnTypeNodes(Demangler &D,
-                                              Node *parent,
-                                              Node *visitedNode) {
-  if (!parent || !visitedNode)
-    return nullptr;
-  if (visitedNode->getKind() == Node::Kind::OpaqueReturnType) {
+/// Implementation for \c setParentForOpaqueReturnTypeNodes. Don't invoke
+/// directly.
+static void setParentForOpaqueReturnTypeNodesImpl(
+    Demangler &D, Node &visitedNode,
+    llvm::function_ref<StringRef()> getParentID) {
+  if (visitedNode.getKind() == Node::Kind::OpaqueReturnType) {
     // If this node is not already parented, parent it.
-    if (visitedNode->hasChildren()
-        && visitedNode->getLastChild()->getKind() == Node::Kind::OpaqueReturnTypeParent) {
-      return parent;
+    if (visitedNode.hasChildren() && visitedNode.getLastChild()->getKind() ==
+                                         Node::Kind::OpaqueReturnTypeParent) {
+      return;
     }
-    visitedNode->addChild(D.createNode(Node::Kind::OpaqueReturnTypeParent,
-                                       (Node::IndexType)parent), D);
-    return parent;
+    visitedNode.addChild(D.createNode(Node::Kind::OpaqueReturnTypeParent,
+                                      StringRef(getParentID())),
```

Comment:
Won't this `StringRef` die when the backing `std::string parentID;` is cleared from the stack when `setParentForOpaqueReturnTypeNodes` returns?

Reply:
No, it won't. There are two types of `createNode` functions:
1) `createNode(Node::Kind K, llvm::StringRef Text)` - Creates a node with `Text` copied
2) `createNodeWithAllocatedText(Node::Kind K, llvm::StringRef Text)`, `createNode(Node::Kind K, const CharVector &Text)`, `createNode(Node::Kind K, const char *Text)` - create a node with `Text` not copied.

I'm intentionally picking the first one here.